### PR TITLE
Hash Join Vectorized collision checking

### LINKED ISSUE #50: Vectorize hash join collision check
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Further optimize the hash join algorithm

**Describe the solution you'd like**
There are a couple of optimizations we could implement:

* Vectorize the row-equality check which now uses the `equal_rows` functions. We should be able to speed this up by vectorizing this, and also specialize it for handling non-null batches too. We probably can utilize the kernels `take` and `eq` here.
* Don't use a `Hashmap` but a `Vec` (or similar) with a certain amount of buckets (proportional to the number of rows or the expected number of keys in the left side). I tried this, but as it causes much more collisions than we have currently, it causes a big (3x) slowdown, so vectorizing the collision check is a prerequisite.

**Additional context**

https://www.cockroachlabs.com/blog/vectorized-hash-joiner/
https://dare.uva.nl/search?identifier=5ccbb60a-38b8-4eeb-858a-e7735dd37487

### PR DESCRIPTION
**Pull request by Dandandan**
# Which issue does this PR close?

Closes #50

# Rationale for this change

Speed up for hash joins:

```

┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃     main ┃ vectorized_collision ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │ 186.15ms │             185.23ms │     no change │
│ QQuery 2     │  68.78ms │              60.46ms │ +1.14x faster │
│ QQuery 3     │  53.50ms │              49.64ms │ +1.08x faster │
│ QQuery 4     │  38.11ms │              38.57ms │     no change │
│ QQuery 5     │ 112.94ms │             101.61ms │ +1.11x faster │
│ QQuery 6     │  11.18ms │              10.97ms │     no change │
│ QQuery 7     │ 228.21ms │             197.74ms │ +1.15x faster │
│ QQuery 8     │  80.80ms │              75.27ms │ +1.07x faster │
│ QQuery 9     │ 166.03ms │             145.26ms │ +1.14x faster │
│ QQuery 10    │ 106.34ms │             100.25ms │ +1.06x faster │
│ QQuery 11    │  55.83ms │              49.45ms │ +1.13x faster │
│ QQuery 12    │  70.98ms │              69.73ms │     no change │
│ QQuery 13    │ 217.69ms │             192.88ms │ +1.13x faster │
│ QQuery 14    │  13.77ms │              12.82ms │ +1.07x faster │
│ QQuery 15    │  23.83ms │              23.92ms │     no change │
│ QQuery 16    │  52.41ms │              53.94ms │     no change │
│ QQuery 17    │ 699.16ms │             699.39ms │     no change │
│ QQuery 18    │ 621.07ms │             547.10ms │ +1.14x faster │
│ QQuery 19    │  61.40ms │              60.98ms │     no change │
│ QQuery 20    │ 218.51ms │             212.73ms │     no change │
│ QQuery 21    │ 341.74ms │             275.30ms │ +1.24x faster │
│ QQuery 22    │  34.32ms │              29.75ms │ +1.15x faster │
└──────────────┴──────────┴──────────────────────┴───────────────┘
```

# What changes are included in this PR?
 Changes the collision checking to be vectorized.

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/joins/hash_join.rs [hunk 1]
```diff
@@ -1097,6 +1098,49 @@ pub fn equal_rows(
...
+        _ => eq_dyn(left, right),
+    }
+}
+
+pub fn equal_rows_arr(
+    indices_left: UInt64Array,
+    indices_right: UInt32Array,
+    left_arrays: &[ArrayRef],
+    right_arrays: &[ArrayRef],
+    null_equals_null: bool,
```
Comment (metesynnada) [inline datafusion/core/src/physical_plan/joins/hash_join.rs:1120]: AFAIK `null_equals_null ` is not supported in the arrow kernel.

File: datafusion/core/src/physical_plan/joins/hash_join.rs [hunk 2]
```diff
@@ -1097,6 +1098,49 @@ pub fn equal_rows(
     err.unwrap_or(Ok(res))
 }
 
+fn eq_dyn_null(
```
Comment (metesynnada) [inline datafusion/core/src/physical_plan/joins/hash_join.rs:1101]: A null element can occur in an `Int32Array` as well. Maybe we should insert an option to arrow kernel, in arrow-rs.

**Comment by metesynnada** [comment]
There is a significant improvement in the equality check, nice work. Maybe then the `apply_join_filter_to_indices` can be addressed, since the intermediate batch can be huge in low cardinality cases 😃

File: datafusion/core/src/physical_plan/joins/hash_join.rs (hunk 1 above)
Comment (Dandandan) [reply datafusion/core/src/physical_plan/joins/hash_join.rs:1120]: Indeed - working on implementing this now

Comment (metesynnada) [reply datafusion/core/src/physical_plan/joins/hash_join.rs:1104]:
```rust
pub fn equal_rows_arr(
    indices_left: UInt64Array,
    indices_right: UInt32Array,
    left_arrays: &[ArrayRef],
    right_arrays: &[ArrayRef],
    null_equals_null: bool,
) -> Result<(UInt64Array, UInt32Array)> {
    // Create an iterator that yields pairs of items from left_arrays and right_arrays.
    let mut iter = left_arrays.iter().zip(right_arrays.iter());

    // Get the first pair of arrays (or error if the iterators are empty).
    let (first_left, first_right) = iter.next().ok_or_else(|| {
        DataFusionError::Internal("At least one array should be provided for both left and right".to_string())
    })?;

    // Use the take function on the first pair of arrays and the respective indices.
    let arr_left = take(first_left.as_ref(), &indices_left, None)?;
    let arr_right = take(first_right.as_ref(), &indices_right, None)?;

    // Determine whether the first pair of arrays are equal, considering nulls as defined by null_equals_null.
    let mut equal = eq_dyn_null(&arr_left, &arr_right, null_equals_null)?;

    // Use map and try_fold to iterate over the remaining pairs of arrays.
    // In each iteration, take is used on the pair of arrays and their equality is determined.
    // The results are then folded (combined) using the and function to get a final equality result.
    equal = iter
        .map(|(left, right)| {
            let arr_left = take(left.as_ref(), &indices_left, None)?;
            let arr_right = take(right.as_ref(), &indices_right, None)?;
            eq_dyn_null(arr_left.as_ref(), arr_right.as_ref(), null_equals_null)
        })
        .try_fold(equal, |acc, res| {
            let equal2 = res?;
            and(&acc, &equal2)
        })?;
    // Continue
}
```

File: datafusion/core/src/physical_plan/joins/hash_join.rs (hunk 2 above)
Comment (metesynnada) [reply datafusion/core/src/physical_plan/joins/hash_join.rs:1077]: I think we can open an issue on arrow-rs for this functionality.

File: datafusion/core/src/physical_plan/joins/hash_join.rs (hunk 1 above)
Comment (metesynnada) [reply datafusion/core/src/physical_plan/joins/hash_join.rs:1104]: In addition, we do not need to take ownership of `indices_left` and `indices_right`.

File: datafusion/core/src/physical_plan/joins/hash_join.rs (hunk 2 above)
Comment (Dandandan) [reply datafusion/core/src/physical_plan/joins/hash_join.rs:1077]: https://github.com/apache/arrow-rs/issues/4438

File: datafusion/core/src/physical_plan/joins/hash_join.rs (hunk 1 above)
Comment (Dandandan) [reply datafusion/core/src/physical_plan/joins/hash_join.rs:1104]: 👍

**Comment by metesynnada** [comment]
LGTM, nice work @Dandandan thanks for the effort.

**Comment by Dandandan** [comment]
> LGTM, nice work @Dandandan thanks for the effort.

Thanks for the feedback!

**Comment by Dandandan** [comment]
Btw I think this can be applied to the symmetric hash join as well without changes.

**Pull request merged**